clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/gitRepo/GitJobExecutor.java

```diff
@@ -0,0 +1,289 @@
+    }
+    if (authType == AuthType.SSH) {
+      return reference.startsWith("ssh://") || reference.startsWith("git@");
+    }
+    return true;
+  }
+
+  private List<String> cmdToList(String cmd) {
+    List<String> cmdList = new ArrayList<>();
+    switch (authType) {
+      case USER_PASS:
+      case TOKEN:
+        // "sh" subshell is used so that environment variables can be used as part of the command
+        cmdList.add("sh");
+        cmdList.add("-c");
+        cmdList.add(cmd);
+        break;
+      case SSH:
+      default:
+        cmdList.addAll(Arrays.asList(cmd.split(" ")));
+        break;
+    }
+    return cmdList;
+  }
+
+  private String repoUrlWithAuth(String repoUrl) {
+    if (authType != AuthType.USER_PASS && authType != AuthType.TOKEN) {
+      return repoUrl;
+    }
+
+    String authPart;
+    if (authType == AuthType.USER_PASS) {
+      authPart = "$GIT_USER:$GIT_PASS";
+    } else {
+      authPart = "token:$GIT_TOKEN";
+    }
+
+    try {
+      URI uri = new URI(repoUrl);
+      return String.format(
+          "%s://%s@%s%s%s",
+          uri.getScheme(),
+          authPart,
+          uri.getHost(),
+          (uri.getPort() > 0 ? ":" + uri.getPort() : ""),
+          uri.getRawPath());
+    } catch (URISyntaxException e) {
+      throw new IllegalArgumentException("Malformed git repo url " + repoUrl, e);
+    }
+  }
+
+  private Map<String, String> addEnvVars(Map<String, String> env) {
+    Map<String, String> result = new HashMap<>(env);
+
+    switch (authType) {
+      case USER_PASS:
+        result.put("GIT_USER", encodeURIComponent(account.getUsername()));
+        result.put("GIT_PASS", encodeURIComponent(account.getPassword()));
+        break;
+      case TOKEN:
+        result.put("GIT_TOKEN", encodeURIComponent(account.getToken()));
+        break;
+      case SSH:
+        result.put("GIT_SSH_COMMAND", buildSshCommand());
+        result.put("SSH_ASKPASS", askPassBinary.toString());
+        result.put("DISPLAY", ":0");
+        if (!StringUtils.isEmpty(account.getSshPrivateKeyPassphrase())) {
+          result.put(SSH_KEY_PWD_ENV_VAR, account.getSshPrivateKeyPassphrase());
+        }
+        break;
+    }
+
+    if (log.isDebugEnabled()) {
+      result.put("GIT_CURL_VERBOSE", "1");
+      result.put("GIT_TRACE", "1");
+    }
+    return result;
+  }
+
+  @NotNull
+  private String buildSshCommand() {
+    String gitSshCmd = "setsid ssh";
+    if (account.isSshTrustUnknownHosts()) {
+      gitSshCmd += " -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no";
+    } else if (!StringUtils.isEmpty(account.getSshKnownHostsFilePath())) {
+      gitSshCmd += " -o UserKnownHostsFile=" + account.getSshKnownHostsFilePath();
+    }
+    if (!StringUtils.isEmpty(account.getSshPrivateKeyFilePath())) {
+      gitSshCmd += " -i " + account.getSshPrivateKeyFilePath();
+    }
+    return gitSshCmd;
+  }
+
+  private static String encodeURIComponent(String s) {
```

Comment:
Out of curiosity, why is this function necessary? Could you give some examples of the expected input/output?

Reply:
Yeah, this is used because of the way the password is sent to git in the username/password authentication method. In that method, the clone url is this:

```
http(s)://$GIT_USER:$GIT_PASS@host:port/contextPath
```

It's possible that the git password contains special characters that could render the url to be invalid, like having `@` or `/`. The method `encodeURIComponent` is used to encode those characters to be sent as part of the url, so instead of 

```
http(s)://john:hds3tr@d/rt@host:port/contextPath
```

It will be 

```
http(s)://john:hds3tr%40d%2Frt@host:port/contextPath
```